app/src/ui/lib/list/section-list-selection.ts

```diff
@@ -0,0 +1,191 @@
+import * as React from 'react'
+import {
+  getTotalRowCount,
+  globalIndexToRowIndexPath,
+  InvalidRowIndexPath,
+  isValidRow,
+  RowIndexPath,
+  rowIndexPathEquals,
+  rowIndexPathToGlobalIndex,
+} from './list-row-index-path'
+
+export type SelectionDirection = 'up' | 'down'
+
+interface ISelectRowAction {
+  /**
+   * The vertical direction use when searching for a selectable row.
+   */
+  readonly direction: SelectionDirection
+
+  /**
+   * The starting row index to search from.
+   */
+  readonly row: RowIndexPath
+
+  /**
+   * A flag to indicate or not to look beyond the last or first
+   * row (depending on direction) such that given the last row and
+   * a downward direction will consider the first row as a
+   * candidate or given the first row and an upward direction
+   * will consider the last row as a candidate.
+   *
+   * Defaults to true if not set.
+   */
+  readonly wrap?: boolean
+}
+
+/**
+ * Interface describing a user initiated selection change event
+ * originating from a pointer device clicking or pressing on an item.
+ */
+export interface IMouseClickSource {
+  readonly kind: 'mouseclick'
+  readonly event: React.MouseEvent<any>
+}
+
+/**
+ * Interface describing a user initiated selection change event
+ * originating from a pointer device hovering over an item.
+ * Only applicable when selectedOnHover is set.
+ */
+export interface IHoverSource {
+  readonly kind: 'hover'
+  readonly event: React.MouseEvent<any>
+}
+
+/**
+ * Interface describing a user initiated selection change event
+ * originating from a keyboard
+ */
+export interface IKeyboardSource {
+  readonly kind: 'keyboard'
+  readonly event: React.KeyboardEvent<any>
+}
+
+/**
+ * Interface describing a user initiated selection of all list
+ * items (usually by clicking the Edit > Select all menu item in
+ * the application window). This is highly specific to GitHub Desktop
+ */
+export interface ISelectAllSource {
+  readonly kind: 'select-all'
+}
+
+/** A type union of possible sources of a selection changed event */
+export type SelectionSource =
+  | IMouseClickSource
+  | IHoverSource
+  | IKeyboardSource
+  | ISelectAllSource
+
+/**
+ * Determine the next selectable row, given the direction and a starting
+ * row index. Whether a row is selectable or not is determined using
+ * the `canSelectRow` function, which defaults to true if not provided.
+ *
+ * Returns null if no row can be selected or if the only selectable row is
+ * identical to the given row parameter.
+ */
+export function findNextSelectableRow(
+  rowCount: ReadonlyArray<number>,
+  action: ISelectRowAction,
+  canSelectRow: (indexPath: RowIndexPath) => boolean = row => true
+): RowIndexPath | null {
+  const totalRowCount = getTotalRowCount(rowCount)
+  if (totalRowCount === 0) {
+    return null
+  }
+
+  const { direction, row } = action
+  const wrap = action.wrap === undefined ? true : action.wrap
```

Comment:
Nit
```suggestion
  const wrap = action.wrap ?? true;
```
`??` looks for `null` or `undefined` not a falsy check like `?`, in case you were wondering if false would be true. :P 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing

Reply:
Yup! That one seems to be old code anyway that I didn't touch, but I can clean it up 😄